lib/spack/spack/fetch_strategy.py

```diff
@@ -339,34 +251,52 @@ class URLFetchStrategy(FetchStrategy):
 
     url_attr = "url"
 
-    # these are checksum types. The generic 'checksum' is deprecated for
-    # specific hash names, but we need it for backward compatibility
-    optional_attrs = [*crypto.hashes.keys(), "checksum"]
-
-    def __init__(self, *, url: str, checksum: Optional[str] = None, **kwargs) -> None:
-        super().__init__(**kwargs)
+    #: Checksum types
+    optional_attrs = [*crypto.hashes.keys()]
 
-        self.url = url
-        self.mirrors = kwargs.get("mirrors", [])
+    def __init__(
+        self,
+        *,
+        url: str,
+        checksum: Optional[str] = None,
+        fetch_options: Optional[FetchOptions] = None,
+        mirrors: Optional[List[str]] = None,
+        expand: bool = True,
+        extension: Optional[str] = None,
+        no_cache: bool = False,
+        **kwargs,
+    ) -> None:
+        super().__init__(no_cache=no_cache)
 
-        # digest can be set as the first argument, or from an explicit
-        # kwarg by the hash name.
+        self.mirrors = mirrors or []
+        # The generic 'checksum' is deprecated for specific hash names,
+        # but we need it for backward compatibility
         self.digest: Optional[str] = checksum
         for h in self.optional_attrs:
             if h in kwargs:
                 self.digest = kwargs[h]
+        self.extension: Optional[str] = extension
+        self.expand_archive = expand
 
-        self.expand_archive: bool = kwargs.get("expand", True)
-        self.extra_options: dict = kwargs.get("fetch_options", {})
-        self._curl: Optional[Executable] = None
-        self.extension: Optional[str] = kwargs.get("extension", None)
-        self._effective_url: Optional[str] = None
+        self.fetch_method, self.fetch_args = default_url_fetch_method()
+        self.timeout, self.cookie = default_timeout(), None
 
-    @property
-    def curl(self) -> Executable:
-        if not self._curl:
-            self._curl = web_util.require_curl()
-        return self._curl
+        # Fetch options are overridden per "resource" being fetched
+        if fetch_options and "timeout" in fetch_options:
+            self.timeout = fetch_options["timeout"]
+
+        if (
+            fetch_options
+            and "cookie" in fetch_options
+            and self.fetch_method == DownloadMethod.CURL
+        ):
+            self.fetch_args.extend(curl_cookie_args(fetch_options["cookie"]))
```

Comment:
This looks like an unsuccessful refactor if the call site has to branch on the fetch method.